### DISCUSSION TITLE
Getting error  OSError("can't create thread",) when try to create a thread inside a thread

### DISCUSSION THREAD

**Discussion by amplatzer** [micropython/micropython#12940] [General]
Micropython 1.20
Module,  ESP32-pico-d4

So my code is like

```
import _thread
def func_a():
    print('func a called')
    
def func_b():
    print('func b called')
    _thread.start_new_thread(func_a,[])
    
_thread.start_new_thread(func_b,[])

```

for the past few weeks, it has worked fine, but today, I found that the second thread except with ERROR **OSError("can't create thread",) .**

What should I do for root cause finding. Can it be something related to memory?
I did increase the size of the .py file in the past few days though.

**Comment by jimmo** [comment]
This error is being returned from the IDF's `xTaskCreatePinnedToCore` function. Perhaps it cannot allocate the stack.

What sort of things have changed in your program since it worekd?

Have you tried in v1.21.0 or v1.22.0-preview?

**Reply by amplatzer** [reply]
What sort of things have changed in your program since it worekd?
—— No really, just increase the size of an unrelieved dictionary. 

Have you tried in v1.21.0 or v1.22.0-preview?
——  No, I was starting this project with v1.21.0, but for some reason, the kernel always crashes when the code try to turn off and on the BLE.

**Reply by amplatzer** [reply]
I do find out, the ERROR always comes out when 2 threads try to do ble.gatts_notify at the same time.  But it doesn't happen every time. So that confuses me.

**Comment by nestorld** [comment]
I'm getting that error (OSError: can't create thread) on a Lilygo T-Beam ESP32 board with Micropython 1.21 (also tried with 1.22). I have several custom libraries running on it for gps, ble and mqtt. All of them use threads and work well when I test them separately, but when I activate all these libraries together I start getting this error. I would say that it is because of scarce resources (mainly memory), but this board has 8 MB of SPIRAM, and it should be more than enough (even with 4 MB as I understand it is the max that ESP32 can use). I have built a custom firmware including these libraries on it to see if that helps, but still getting the same error. I did use this command to build the firmware:

make BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM FROZEN_MANIFEST="" FROZEN_MPY_DIR=$UPYDIR/modules

Something I'm not sure about is if the code is really making use of the SPIRAM. I have used micropython.mem_info() in the Thonny's REPL console and got this:

stack: 736 out of 15360
GC: total: 64000, used: 15824, free: 48176, max new split: 4063232
 No. of 1-blocks: 356, 2-blocks: 81, max blk sz: 32, max free sz: 2999

Am I missing something to give enough resources to my code to use these threads?
Thanks in advance for any help on this!

**Reply by amplatzer** [reply]
```py
import _thread
from micropython import mem_info
from gc import mem_free,collect
import time

def test():
    time.sleep(30)
    
try:
    for i in range(30):
        print(i)
        _thread.start_new_thread(test,[])
        time.sleep(0.5)
except:
    print('MEM_FREE: ',mem_free())
    print('########')
    print(mem_info())

```

IMicroPython v1.20.0-dirty on 2023-11-07; ESP32 module with ESP32
Type "help()" for more information.
%run -c $EDITOR_CONTENT
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
MEM_FREE:  98992
########
stack: 720 out of 15360
GC: total: 108032, used: 9040, free: 98992
 No. of 1-blocks: 187, 2-blocks: 51, max blk sz: 32, max free sz: 6187
None



I'm using esp32-pico-d4, looks like there is a limitation on thread? when it    comes to my code, I can only create 2 threads, and getting error  with the 3rd one. But what can we learn from the memery info ?\

**Reply by amplatzer** [reply]
maybe you can try _thread.stack_size(4096) at the begging of your code.
